Validate if `utf8mb4` character set and longer index key prefix is supported

yahonda opened the pull request:
### Summary

Use `utf8` character set by default if MySQL server does not support longer index key prefix

Once #33608 merged If users create a new database using MySQL 5.1.x, it will fail to create databases
since MySQL 5.1 does not know `utf8mb4` character set.

This pull request removes `encoding: utf8mb4` from `mysql.yml.tt`
to let create_database method handles default character set by MySQL server version.

### Other Information

`supports_longer_index_key_prefix?` method will need to validate if MySQL 5.5 and 5.6 server configured correctly to support longer index key prefix, but not yet.

yahonda committed:
Validate if `utf8mb4` character set and longer index key prefix is supported

Once #33608 merged If users create a new database using MySQL 5.1.x, it will fail to create databases
since MySQL 5.1 does not know `utf8mb4` character set.

This pull request removes `encoding: utf8mb4` from `mysql.yml.tt`
to let create_database method handles default character set by MySQL server version.

`supports_longer_index_key_prefix?` method will need to validate if MySQL 5.5 and 5.6 server configured
correctly to support longer index key prefix, but not yet.

jeremy commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -114,6 +114,14 @@ def supports_advisory_locks?
+        else
+          version >= "5.7.9"
+        end
+      end
```
Should we feature-detect `innodb_large_prefix` instead of version check?

jeremy commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -250,8 +258,12 @@ def recreate_database(name, options = {})
+          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET `utf8mb4`"
         else
-          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET #{quote_table_name(options[:charset] || 'utf8mb4')}"
+          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET `utf8`"
```
Perhaps we can safely raise an exception now. If `:charset` is not specified and large prefixes / utf8mb4 are not supported, we don't want to silently default to 3-byte anymore.

yahonda committed:
Raise an exception if :charset is not specified and large prefixes / utf8mb4 are not supported

kamipo commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -250,8 +258,12 @@ def recreate_database(name, options = {})
+          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET `utf8mb4`"
         else
-          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET #{quote_table_name(options[:charset] || 'utf8mb4')}"
+          raise "Your MySQL (#{version_string}) does not support utf8mb4 character and longer index key prefix. Configure `:charset` in your database.yml"
```
How about dropping MySQL 5.1 support?
MySQL 5.1 is only version that ``CREATE DATABASE `table_name` DEFAULT CHARACTER SET `utf8mb4` `` is failed which mysql2 adapter supported.

yahonda committed:
Drop MySQL 5.1 support

* MySQL 5.1 does not support `utf8mb4` character set

* MySQL 5.1 had been already EOLed on Dec 2013
https://www.mysql.com/support/eol-notice.html

> Per Oracle's Lifetime Support policy, as of December 31, 2013, MySQL 5.1
> is covered under Oracle Sustaining Support.

* MySQL 5.5.8 is the first General Availability of MySQL 5.5

https://dev.mysql.com/doc/relnotes/mysql/5.5/en/news-5-5-8.html

kamipo commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -250,8 +258,12 @@ def recreate_database(name, options = {})
+          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET `utf8mb4`"
         else
-          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET #{quote_table_name(options[:charset] || 'utf8mb4')}"
+          raise "Configure a supported :charset in config/database.yml and ensure innodb_large_prefix is enabled to support indexes on varchar(255) string columns."
```
Should we mention about `config/database.yml` in this method?
`create_database` could work without `config/database.yml`.

yahonda replied:
I see. Removed `config/database.yml` from the message.
